webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala

```diff
@@ -262,32 +287,91 @@ trait CachedFindRuleNodeStatusReports extends ReportingService with CachedReposi
-        _       <- IOResult.effectNonBlocking { cache = cache ++ updated }
-        _       <- ReportLoggerPure.Cache.debug(s"Compliance cache updated for nodes: ${nodeIds.map(_.value).mkString(", ")}")
-      } yield ()).catchAll(err => ReportLoggerPure.Cache.error(s"Error when updating compliance cache for nodes: [${nodeIds.map(_.value).mkString(", ")}]: ${err.fullMsg}"))
+    ZIO.foreach_(groupQueueActionByType(invalidatedIds.map(x => x._2)).to(Iterable))(actions =>
+      // several strategy:
+      // * we have a compliance: yeah, put it in the cache
+      // * new policy generation, a new nodeexpectedreports is available: compute compliance for last run of the node, based on this nodeexpectedreports
+      // * node deletion: remove from the cache
+      // * invalidate cache: ???
+      // * no report from the node (compliance expires): recompute compliance
+
+      // as a first approach, they could simply findRuleNodeStatusReports
+      {
+        (for {
+          _  <- performAction(actions)
+        } yield ()).catchAll(err => ReportLoggerPure.Cache.error(s"Error when updating compliance cache for nodes: [${actions.map(_.nodeId).map(_.value).mkString(", ")}]: ${err.fullMsg}"))
+      }
     )
   )
 
   // start updating
   updateCacheFromRequest.forever.forkDaemon.runNow
 
 
+  /**
+   * Do something with the action we received
+   * All actions must have *exactly* the *same* type
+   */
+  private[this] def performAction(actions: List[CacheComplianceQueueAction]): IOResult[Unit] = {
+    ReportLoggerPure.Cache.debug(s"Performing action ${actions.headOption}") *>
+    // get type of action
+      (actions.headOption match {
+      case None => ReportLoggerPure.Cache.debug("Nothing to do")
+      case Some(t) => t match {
+        case update:UpdateCompliance =>
+          ReportLoggerPure.Cache.debug(s"Compliance cache updated for nodes: ${actions.map(_.nodeId.value).mkString(", ")}") *>
+          IOResult.effectNonBlocking {
+            cache = cache ++ actions.map { case x: UpdateCompliance => (x.nodeId, x.nodeCompliance) }
+          }
+        case delete: RemoveNodeInCache =>
+          IOResult.effectNonBlocking {
+            cache = cache.removedAll(actions.map { case x: RemoveNodeInCache => (x.nodeId) })
+          }
+
+        // need to compute compliance
+        case _ =>
+          val impactedNodeIds = actions.map(x => x.nodeId)
+          for {
+              x <- ZIO.foreach(impactedNodeIds.grouped(batchSize).to(Seq)) { updatedNodes =>
+                  for {
+                  updated <- defaultFindRuleNodeStatusReports.findRuleNodeStatusReports(updatedNodes.toSet, Set()).toIO
+                  _       <- IOResult.effectNonBlocking {
+                  cache = cache ++ updated
+                }
+                _ <- ReportLoggerPure.Cache.debug(s"Compliance cache recomputed for nodes: ${updated.keys.map(_.value).mkString(", ")}")
+                } yield ()
+            }
+          } yield ()
```

Comment:
bad indent for `for`, shoud look like: 

```
              x <- ZIO.foreach(impactedNodeIds.grouped(batchSize).to(Seq)) { updatedNodes =>
                       for {
                         updated <- defaultFindRuleNodeStatusReports.findRuleNodeStatusReports(updatedNodes.toSet, Set()).toIO
                         _       <- IOResult.effectNonBlocking {
                                      cache = cache ++ updated
                                    }
etc
```
(can even be `IOResult.effectNonBlocking { cache = cache ++ updated }`)